pylops/signalprocessing/dtcwt.py

```diff
@@ -0,0 +1,145 @@
+    birot: :obj:`str`, optional
+        Level 1 wavelets to use. See :py:func:`dtcwt.coeffs.birot()`. Default is `"near_sym_a"`.
+    qshift: :obj:`str`, optional
+        Level >= 2 wavelets to use. See :py:func:`dtcwt.coeffs.qshift()`. Default is `"qshift_a"`
+    nlevels: :obj:`int`, optional
+        Number of levels of wavelet decomposition. Default is 3.
+    include_scale: :obj:`bool`, optional
+        Include scales in pyramid. See :py:func:`dtcwt.Pyramid`. Default is False.
+    dtype : :obj:`DTypeLike`, optional
+        Type of elements in input array.
+    name : :obj:`str`, optional
+        Name of operator (to be used by :func:`pylops.utils.describe.describe`)
+
+    Raises
+    ------
+    NotImplementedError
+        If ``transform`` is 2 or 3.
+    ValueError
+        If ``transform`` is anything other than 1, 2 or 3.
+
+    Notes
+    -----
+    The :py:func:`dtcwt` library uses a Pyramid object to represent the transform domain signal.
+    It has
+        `lowpass` (coarsest scale lowpass signal)
+        `highpasses` (complex subband coefficients for corresponding scales)
+        `scales` (lowpass signal for corresponding scales finest to coarsest)
+
+    To make the dtcwt forward() and inverse() functions compatible with pylops, the Pyramid object is
+    flattened out and all coeffs(highpasses and low pass signal) are appened into one array using the
+    `_coeff_to_array` method.
+    For inverse, the flattened array is used to reconstruct the Pyramid object using the `_array_to_coeff`
+    method and then inverse is performed.
+
+    """
+
+    def __init__(
+        self,
+        dims: Union[int, InputDimsLike],
+        transform: int = 1,
+        biort: str = "near_sym_a",
+        qshift: str = "qshift_a",
+        nlevels: int = 3,
+        include_scale: bool = False,
+        dtype: DTypeLike = "float64",
+        name: str = "C",
+    ) -> None:
+        self.dims = _value_or_sized_to_tuple(dims)
+        self.transform = transform
+        self.ndim = len(self.dims)
+        self.nlevels = nlevels
+        self.include_scale = include_scale
+
+        if self.transform == 1:
+            self._transform = dtcwt.Transform1d(biort=biort, qshift=qshift)
+        elif self.transform == 2:
+            raise NotImplementedError("DTCWT is not implmented for 2D")
+        elif self.transform == 3:
+            raise NotImplementedError("DTCWT is not implmented for 3D")
+        else:
+            raise ValueError("DTCWT only supports 1D, 2D and 3D")
+
+        pyr = self._transform.forward(
+            np.ones(self.dims), nlevels=self.nlevels, include_scale=True
+        )
+        self.coeff_array_size = 0
+        self.lowpass_size = len(pyr.lowpass)
+        self.slices = []
+        for _h in pyr.highpasses:
+            self.slices.append(len(_h))
+            self.coeff_array_size += len(_h)
+        self.coeff_array_size += self.lowpass_size
+        self.second_dim = 1
+        if len(dims) > 1:
```

Comment:
What happens if dims has size of 3. From the documentation of `dtcwt.Transform1d` it seems they only allow handling 1d or 2d arrays (their doc: `X – 1D real array or 2D real array whose columns are to be transformed`)... but we should allow also 3d (or any nd) and also allow applying the transform on any axis.

This can be easily done by combining:

- adding `axis` to inputs with default to -1. Note that dtcwt applies the transform over columns so we are good :)
- `@reshaped(swapaxis=True)` see CausalIntegration for an example.. basically the axis over which to apply the transform is moved to last
- reshape x before feeding it to the transform such that we have a 2D array `otherdims * dims[axis]` (and reshape back to nd array in adjoint)

Reply:
Also since they say the transform is applied over columns, why do you have second_dim and not first_dim as the one that broadcasts?

Reply:
I made two funcs that converts the dims to 2D and back to original shape.
[f887031](https://github.com/PyLops/pylops/pull/495/commits/f887031ff0aed0edac70bed9bf1388e2dbfb4b70)